### PR TITLE
Fix/2990/label-height-issue

### PR DESCRIPTION
**Pull request by RomanenkoVladimir**
# Fix label height issue

_**Please read the [CONTRIBUTING.md](https://github.com/MaibornWolff/codecharta/blob/main/CONTRIBUTING.md) before opening a PR.**_

closes: #2990

## Description

Fixes various height issues for leaf labels when changing the height scaling.

### CODE REVIEW

File: visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts [hunk 1]
```diff
@@ -29,6 +29,7 @@ export class CodeMapLabelService implements CameraChangeSubscriber {
 	private previousScaling: Vector3 = new Vector3(1, 1, 1)
 	private lineCount = 1
 	private nodeHeight = 0
+	private HEIGHT_OFFSET = 1
```
Comment (knoffi) [inline visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts:32]: I would expect an offset to be added to something, not multiplied. How about `HEIGHT_BASE_FACTOR` instead?

**Review by knoffi — DISMISSED** [review]
The unit-tests are failing. Please check my comments for this, a mock object has to be updated.

File: visualization/app/codeCharta/ui/codeMap/codeMap.render.service.spec.ts [hunk 2]
```diff
@@ -246,7 +246,7 @@ describe("codeMapRenderService", () => {
 			codeMapRenderService["getNodes"] = jest.fn().mockReturnValue(nodes)
 			codeMapRenderService.render(null)
 
-			expect(codeMapLabelService.addLabel).toHaveBeenCalledTimes(1)
+			expect(codeMapLabelService.addLeafLabel).toHaveBeenCalledTimes(1)
```
[the remark this answers is not in the document]
Comment (knoffi) [reply visualization/app/codeCharta/ui/codeMap/codeMap.render.service.spec.ts:249]: Resolved with commit [592ff26](https://github.com/MaibornWolff/codecharta/pull/2991/commits/592ff26c64a34b61ea430b50be9d7eeb7d42bf31)

File: visualization/app/codeCharta/ui/codeMap/codeMap.render.service.spec.ts [hunk 3]
```diff
@@ -216,7 +216,7 @@ describe("codeMapRenderService", () => {
 			codeMapRenderService["setLabels"]([])
 
 			expect(codeMapLabelService.clearLabels).toHaveBeenCalled()
-			expect(codeMapLabelService.addLabel).not.toHaveBeenCalled()
+			expect(codeMapLabelService.addLeafLabel).not.toHaveBeenCalled()
```
[the remark this answers is not in the document]
Comment (knoffi) [reply visualization/app/codeCharta/ui/codeMap/codeMap.render.service.spec.ts:219]: Resolved with commit [592ff26](https://github.com/MaibornWolff/codecharta/pull/2991/commits/592ff26c64a34b61ea430b50be9d7eeb7d42bf31)

File: visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts (hunk 1 above)
Comment (BridgeAR) [reply visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts:32]: I agree while it seems to only be read once and never changed. Therefore, we could just remove the variable again.

Comment (RomanenkoVladimir) [reply visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts:32]: Yes, I will push a changed version later today.
I tried something out and this is a remainder from it

File: visualization/package.json [hunk 4]
```diff
@@ -45,7 +45,7 @@
 	"scripts": {
 		"start": "nw .",
 		"build": "webpack --config ./conf/webpack.config.js --env DEV=false",
-		"dev": "webpack serve --config ./conf/webpack.config.js --env DEV=true",
+		"dev": "NODE_OPTIONS=--openssl-legacy-provider webpack serve --config ./conf/webpack.config.js --env DEV=true",
```
Comment (Hall-Ma) [inline visualization/package.json:48]: Do we need this option? I cannot run the dev mode anymore.

File: visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts [hunk 5]
```diff
@@ -169,16 +168,21 @@ export class CodeMapLabelService {
 		const { scaling } = this.storeService.getState().appSettings
 		const { margin } = this.storeService.getState().dynamicSettings
 
-		const multiplier = scaling.clone().divide(this.previousScaling)
+		const labelHeightDifference = new Vector3(0, this.LABEL_HEIGHT_COEFFICIENT * margin * this.LABEL_SCALE_FACTOR, 0)
 
 		for (const label of this.labels) {
-			const labelHeightDifference = new Vector3(0, this.LABEL_HEIGHT_COEFFICIENT * margin * this.LABEL_SCALE_FACTOR, 0)
-			label.sprite.position.sub(labelHeightDifference).multiply(multiplier).add(labelHeightDifference)
+			const multiplier = scaling.clone()
```
Comment (Hall-Ma) [inline visualization/app/codeCharta/ui/codeMap/codeMap.label.service.ts:174]: Please see also line 49. Why do we have two times `const multiplier = scaling.clone()`?

File: visualization/package.json (hunk 4 above)
Comment (RomanenkoVladimir) [reply visualization/package.json]: It is needed to run the project using 18.7.0 but given that the recommended version is 16 I removed the change

**Pull request merged**